jena-arq/src/main/java/org/apache/jena/sparql/engine/join/MultiHashProbeTable.java

```diff
@@ -0,0 +1,196 @@
+        // so that those correspond to the first bits of the bit keys
+        seenVarSet.addAll(initialJoinKey);
+
+        int nbits = initialJoinKey.size();
+        BitSet initialJoinKeyBitset = new BitSet(nbits);
+        initialJoinKeyBitset.flip(0, nbits);
+
+        if (logger.isTraceEnabled()) {
+            logger.trace("Initial join index configured with variables " + initialJoinKey + " and bits " + initialJoinKeyBitset);
+        }
+
+        this.initialIndex = new JoinIndex(initialJoinKey, initialJoinKey, initialJoinKeyBitset);
+    }
+
+    public void put(Binding row) {
+        if (isFinalized) {
+            throw new IllegalStateException("Cannot add more bindings after a lookup was performed.");
+        }
+        updateSeenVars(row);
+        initialIndex.put(row);
+    }
+
+    /** Update seen vars with the row's relevant variables w.r.t. an optional rootJoinKey. */
+    private void updateSeenVars(Binding row) {
+        if (maxJoinKey == null) {
+            for (Iterator<Var> it = row.vars(); it.hasNext(); ) {
+                Var var = it.next();
+                seenVarSet.add(var);
+            }
+        } else {
+            // Iterate over the smallest set of variables
+            if (maxJoinKey.length() < row.size()) { // join key has fewer vars than the row
+                for (Iterator<Var> it = maxJoinKey.iterator(); it.hasNext(); ) {
+                    Var var = it.next();
+                    if (row.contains(var)) {
+                        seenVarSet.add(var);
+                    }
+                }
+            } else { // row has fewer vars than the join key
+                for (Iterator<Var> it = row.vars(); it.hasNext(); ) {
+                    Var var = it.next();
+                    if (maxJoinKey.contains(var)) {
+                        seenVarSet.add(var);
+                    }
+                }
+            }
+        }
+    }
+
+    public Iterator<Binding> getCandidates(Binding row) {
+        if (!isFinalized) {
+            doFinalize();
+        }
+        BitSet joinKeyBitSet = BitSetMapper.toBitSet(seenVarsJoinKey, row);
+        if (isDebugOutputEnabled) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("Lookup with " + BitSetMapper.toList(seenVarsJoinKey, joinKeyBitSet));
+            }
+        }
+
+        Iterator<Binding> it;
+        if (joinKeyBitSet.isEmpty()) {
+            // Case for no joining variables: all bindings unconditionally become candidates
+            it = initialIndex.iterator();
+        } else {
+            JoinIndex primaryIndex = getOrCreateJoinIndex(joinKeyBitSet);
+            it = primaryIndex.getCandidates(row);
+        }
+        return it;
+    }
+
+    /**
+     * Calling this method indicates that all bindings have been collected and are ready for indexing.
+     */
+    // XXX Instead of doFinalize we could introduce a MultiHashProbeTableBuilder that returns an immutable HashProbeTable.
```

Comment:
Please remove, or decide to have MultiHashProbeTableBuilder.

Reply:
Removed comment.